File: 프로그래머스/3/43238. 입국심사/입국심사.java
import java.util.Arrays;

class Solution {
    public long solution(int n, int[] times) {
        long left = 0;
        long right = (long) Arrays.stream(times).max().getAsInt() * n;
        
        long answer = right;
        
        while (left <= right) {
            long mid = (left + right) / 2;
            
            long total = 0;
            for (int t : times) {
                total += mid / t;
            }
            
            if (total >= n) {
                answer = mid;
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }        
        
        return answer;
    }
}

File: 프로그래머스/3/43238. 입국심사/입국심사.py
def solution(n, times):
    times.sort()
    answer = times[-1] * n
    
    min_time = 0
    max_time = answer
    
    while min_time < max_time:
        current = (min_time + max_time) // 2
        
        required = 0
        for t in times:
            required += current // t
            if required >= n:
                break
        
        if required >= n:
            answer = min(answer, current)
            max_time = current
        else:
            min_time = current + 1
    
    return answer
